Assignment03/step-by-step-python.md

```diff
@@ -0,0 +1,370 @@
+   runs on HTTP port `3600`. Replace the URL in the HTTP call with a call to the Dapr pub/sub API:
+
+   ```shell
+   FINE_COLLECTION_ADDRESS=http://localhost:3600/v1.0/publish/pubsub
+   ```
+
+That's it. You now use Dapr pub/sub to publish a message to a message broker.
+
+## Step 4: Receive messages in the FineCollectionService (declaratively)
+
+Now you are going to prepare the FineCollectionService so it can receive messages using Dapr pub/sub. Consuming
+messages can be done in two ways: *declaratively* (through configuration) or *programmatic* (from the code). First
+you'll use the declarative way. Later you'll also use the programmatic way and finally also using the Dapr SDK for
+Python.
+
+1. Add a new file in the `python/dapr/components` folder named `subscription.yaml`.
+
+2. Open this file in VS Code.
+
+3. You're going to define a subscription on a topic and link that to a web API operation on the FineCollectionService.
+   Paste this snippet into the file:
+
+   ```yaml
+   apiVersion: dapr.io/v1alpha1
+   kind: Subscription
+   metadata:
+     name: speedingviolations-subscription
+   spec:
+     topic: speedingviolations
+     route: /collectfine
+     pubsubname: pubsub
+   scopes:
+   - finecollectionservice
+   ```
+
+   The `route` field tells Dapr to forward all messages send to the `speedingviolations` topic to the `/collectfine`
+   endpoint in the app. The `scopes` field restricts this subscription to only the service with app-id
+   `finecollectionservice` only.
+
+Now your FineCollectionService is ready to receive messages through Dapr pub/sub. But there is a catch! Dapr uses the
+[CloudEvents](https://cloudevents.io/) message format for pub/sub. So when we send a message through pub/sub, the
+receiving application needs to understand this format and handle the message as a `CloudEvent`. Therefore we need to
+change the code slightly. For now we write some code, later we're going to use the Dapr SDK.
+
+1. Open the file `python\FineCollectionService\fine_collection\__init__.py` in VS Code.
+
+2. Remove the `violation: models.SpeedingViolation` parameter and replace it with `evt_data=Body(...)`.
+
+   After replacing the parameter you should have a method definition that looks 
+   like this:
+
+   ```python
+   def collect_fine(evt_data=Body(...)) -> Response:
+   ```
+
+   You now have access to the incoming HTTP request data including the body of the request that contains the event
+   that's sent to the FineCollectionService.
+
+3. Add the following code to the top of the body of the `collect_fine` method to extract the `SpeedingViolation` data
+   from the request:
+
+   ```python
+   violation = models.SpeedingViolation.parse_raw(evt_data["data"])
+   ```
+
+4. Add the following import statement to the top of the file to get the `Body` class.
+
+   ```python
+   from fastapi import Body
+   ```
+
+5. Save the changes to the file.
```

Comment:
I think I followed these instructions to the letter. When I run the simulation, the logging of the Traffic Control Service shows

```console
== APP == INFO:     127.0.0.1:34746 - "POST /entrycam HTTP/1.1" 200 OK
== APP == INFO:     127.0.0.1:34750 - "POST /exitcam HTTP/1.1" 200 OK
== APP == Vehicle XT-346-Y is over the speed limit. Collecting fine
== APP == Received status 204 from finecollection service: b''
```

I think this is because later in the `collect_fine` of the `FineCollectionClient`, there is a check on the status of the HTTP call response. Invoking the Fine Collection Service directly results in status 200 (and the code checks for that) but invoking the pub/sub building block on the Dapr sidecar results in status 204. As a result, the check fails, and a `ClientError` is raised.

Reply:
Shouldn't the code be changed to post the violations to the correct pub/sub topic? I believe the code now posts to the **collectfine** topic. Running the simulation does not make the Fine Collection Service log anything, even though there are plenty speeding violations.

If my suspicion is correct, the `requests.post` invocation needs to change so the message is posted to the **speedingviolations** topic. When I do that, the Fine Collection Service starts spitting out these snacktraces:

```console
ERRO[0260] rabbitmq pub/sub: error handling message from topic 'speedingviolations', retriable error returned from app while processing pub/sub event 87303d21-6058-4b50-af5d-bd49c2f67a90: Internal Server Error. status code returned: 500  app_id=finecollectionservice instance=edsger scope=dapr.contrib type=log ver=1.2.2
== APP == ERROR:    Exception in ASGI application
== APP == Traceback (most recent call last):
== APP ==   File "/home/maarten/.local/lib/python3.9/site-packages/uvicorn/protocols/http/h11_impl.py", line 373, in run_asgi
== APP ==     result = await app(self.scope, self.receive, self.send)
== APP ==   File "/home/maarten/.local/lib/python3.9/site-packages/uvicorn/middleware/proxy_headers.py", line 75, in __call__
== APP ==     return await self.app(scope, receive, send)
== APP ==   File "/home/maarten/.local/lib/python3.9/site-packages/fastapi/applications.py", line 208, in __call__
== APP ==     await super().__call__(scope, receive, send)
== APP ==   File "/home/maarten/.local/lib/python3.9/site-packages/starlette/applications.py", line 112, in __call__
== APP ==     await self.middleware_stack(scope, receive, send)
== APP ==   File "/home/maarten/.local/lib/python3.9/site-packages/starlette/middleware/errors.py", line 181, in __call__
== APP ==     raise exc from None
== APP ==   File "/home/maarten/.local/lib/python3.9/site-packages/starlette/middleware/errors.py", line 159, in __call__
== APP ==     await self.app(scope, receive, _send)
== APP ==   File "/home/maarten/.local/lib/python3.9/site-packages/starlette/exceptions.py", line 82, in __call__
== APP ==     raise exc from None
== APP ==   File "/home/maarten/.local/lib/python3.9/site-packages/starlette/exceptions.py", line 71, in __call__
== APP ==     await self.app(scope, receive, sender)
== APP ==   File "/home/maarten/.local/lib/python3.9/site-packages/starlette/routing.py", line 580, in __call__
== APP ==     await route.handle(scope, receive, send)
== APP ==   File "/home/maarten/.local/lib/python3.9/site-packages/starlette/routing.py", line 241, in handle
== APP ==     await self.app(scope, receive, send)
== APP ==   File "/home/maarten/.local/lib/python3.9/site-packages/starlette/routing.py", line 52, in app
== APP ==     response = await func(request)
== APP ==   File "/home/maarten/.local/lib/python3.9/site-packages/fastapi/routing.py", line 226, in app
== APP ==     raw_response = await run_endpoint_function(
== APP ==   File "/home/maarten/.local/lib/python3.9/site-packages/fastapi/routing.py", line 161, in run_endpoint_function
== APP ==     return await run_in_threadpool(dependant.call, **values)
== APP ==   File "/home/maarten/.local/lib/python3.9/site-packages/starlette/concurrency.py", line 40, in run_in_threadpool
== APP ==     return await loop.run_in_executor(None, func, *args)
== APP ==   File "/usr/lib/python3.9/concurrent/futures/thread.py", line 52, in run
== APP ==     result = self.fn(*self.args, **self.kwargs)
== APP ==   File "/home/maarten/Code/dapr-workshop-python/python/FineCollectionService/./fine_collection/__init__.py", line 21, in collect_fine
== APP ==     violation = models.SpeedingViolation.parse_raw(evt_data["data"])
== APP ==   File "pydantic/main.py", line 600, in pydantic.main.BaseModel.parse_raw
== APP == pydantic.error_wrappers.ValidationError: 1 validation error for SpeedingViolation
== APP == __root__
== APP ==   the JSON object must be str, bytes or bytearray, not dict (type=type_error)
```

Reply:
The strange thing is that when I update the Traffic Control Service to use the Dapr SDK (last step of this assignment), there is no snacktrace and the call seems to succeed.

Reply:
I think I know why. I had an async method before. But that caused a problem and I choose to replace it with a normal method definition for the API handler. I didn't check the original HTTP version ;)

Reply:
Nah, it doesn't. I need to remove a bit of code here.

Reply:
Allrighty, this is now working as expected.